### ISSUE TITLE
Inconsistent handling of arguments for disposable types

### ISSUE THREAD

**Issue by psfinaki** [dotnet/fsharp#15622]
This code produces an error only in the second case:
```fsharp
open System.IO

let stream1 = new MemoryStream 42

let capacity = 42
let stream2 = new MemoryStream capacity
```
![image](https://github.com/dotnet/fsharp/assets/5451366/2a60cf5e-7895-425d-9b2e-f3bd8edb8391)

Note there is no error if `new` is unused (there are legitimate warnings though):

![image](https://github.com/dotnet/fsharp/assets/5451366/80ac051d-ac27-4528-aa51-39b6786f6dbc)

We need to decide what to do about that before addressing [this ticket](https://github.com/dotnet/fsharp/issues/6381) - basically either fix the parser or the code fix.

**Comment by vzarytovskii** [comment]
We should just stop recommending new for disposables.

**Comment by psfinaki** [comment]
> We should just stop recommending new for disposables.

Well this is sort of orthogonal. We may remove the code fix, or the diagnostic together with the code fix, but as long as the syntax in question is allowed, the inconsistency will stay.

**Comment by edgarfgp** [comment]
@brianrourkeboll Might this also be fixed by https://github.com/dotnet/fsharp/pull/16239 ?

**Comment by brianrourkeboll** [comment]
> @brianrourkeboll Might this also be fixed by #16239 ?

That PR just updated the analyzer and code fix for removing unneeded parens so that it does _not_ suggest removing parens in such cases.

The solution I originally wondered about

> Is the `atomicExprAfterType` restriction here actually still needed? Would that suddenly cause more problems like those addressed by https://github.com/dotnet/fsharp/pull/15923? Or could we just replace it with `argExpr` in the parser for `new` and `inherit` and get rid of the inconsistency? Making that change locally seems to work and doesn't seem to cause any immediate problems in this codebase...

is unfortunately a breaking change, as shown by the tests linked in https://github.com/dotnet/fsharp/pull/16239#issuecomment-1801860477.

But yes, the problem in @psfinaki's issue here is indeed the same underlying thing.

**Comment by brianrourkeboll** [comment]
We _could_ probably add an analyzer and code fix that offers to add parentheses in cases like this, though.

**Comment by psfinaki** [comment]
I can create a ticket for that - but first let me just clarify, do I understand it correctly that we cannot do anything in the compiler about fixing the root cause? As in, that would be an undesirable breaking change?

**Comment by brianrourkeboll** [comment]
> I can create a ticket for that - but first let me just clarify, do I understand it correctly that we cannot do anything in the compiler about fixing the root cause? As in, that would be an undesirable breaking change?

Yes, unfortunately. Making that change would cause code using postfix generic notation after `new` or `inherit` no longer compile:

https://github.com/dotnet/fsharp/blob/24ef67194b4fa3366e2e87fc77a76c068ce78889/tests/fsharpqa/Source/Conformance/Expressions/ApplicationExpressions/BasicApplication/E_PostfixType01.fs#L6-L7

https://github.com/dotnet/fsharp/blob/24ef67194b4fa3366e2e87fc77a76c068ce78889/tests/FSharp.Compiler.ComponentTests/Conformance/Expressions/ApplicationExpressions/BasicApplication/E_PostfixType03.fs#L6-L10

Those bits of code above no longer compiled after I updated the parser locally—the parens were then being parsed as part of the _type_, not as the argument to the constructor.

It's a shame we can't change it, because I definitely run into the inconsistency far more often than I want to use postfix notation in such scenarios (I don't think I ever have).

**Comment by brianrourkeboll** [comment]
...Although I do wonder whether it might be possible to add new parsing for independent use of `new` alone—i.e., just `new T ()`, not as part of an object expression `{ new T () with … }`—while being backwards-compatible.

Basically: what if we changed the parsing in that case so that everything between the `new` and the last atomic expression was parsed as the type? We know in the independent scenario that exactly one constructor argument must be provided.

The only reason that the parser currently has this oddity is because in `new` expressions inside of object expressions, or in `inherit` expressions, you can have types _without_ constructor arguments, e.g., abstract classes or interfaces, and the parsing code is shared between that use of `new` and independent `new`—but, as above, independent `new` always requires exactly one constructor argument, so its parsing could be different.

**Comment by psfinaki** [comment]
Hah, that's what I call a fresh thought :D 

@auduchinok what do you think?

**Comment by auduchinok** [comment]
> Basically: what if we changed the parsing in that case so that everything between the new and the last atomic expression was parsed as the type? We know in the independent scenario that exactly one constructor argument must be provided.

> @auduchinok what do you think?

That's a interesting idea! From the top of my head it seems it _might_ work. A somewhat similar trick is used in record expression parsing, where `F = 1` inside `{ F = 1 }` is initially parsed as a binary expression and then converted to a field binding by the parser.

A possible problem is this: right now the parser can parse various (not allowed) type usages and the type checker gives an error, while if we try to parse the things as expressions, some of the types might not be parsed anymore. And all the types have to be parsed properly during typing even until the actual arg expression is present.

Another point to think about is we'd need to update features code for the new tree structure. Consider this:
```fsharp
new T {caret}
```
`{caret}` is where the IDE cursor is. With such a change, what's going to be written at that location could be both a type and an expression, but will be parsed as an expression. So code completion has to provide all the available type names too on such a node.

In any case, if we agree on such a change, it should probably have a really good test coverage. 🙂

**Comment by psfinaki** [comment]
Fixed by https://github.com/dotnet/fsharp/pull/18179, closing 🚀
